### PR TITLE
RAC-257: fix accessibility for existing components

### PR DESCRIPTION
**Pull request by juliensnz**
**Description (for Contributor and Core Developer)**

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Todo
| Added legacy Behats               | Todo
| Added acceptance tests            | Todo
| Added integration tests           | Todo
| Changelog updated                 | Todo
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: akeneo-design-system/package.json [hunk 1]
```diff
@@ -40,6 +40,7 @@
     "@babel/core": "^7.3.4",
     "@babel/plugin-syntax-flow": "^7.10.4",
     "@rollup/plugin-replace": "^2.3.3",
+    "@storybook/addon-a11y": "^6.0.21",
```
Comment (juliensnz) [inline akeneo-design-system/package.json:43]: I don't really know if this is necessary as it doesn't really give good informations: only contrast problems (and there is a lot of them)

File: akeneo-design-system/src/components/Button/Button.tsx [hunk 2]
```diff
@@ -137,7 +137,7 @@ const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
     }: ButtonProps,
     forwardedRef: Ref<HTMLButtonElement>
   ) => {
-    useShortcut(Key.Enter, disabled ? () => null : onClick);
+    const ref = useShortcut(Key.Enter, disabled ? () => null : onClick, forwardedRef);
```
Comment (juliensnz) [inline akeneo-design-system/src/components/Button/Button.tsx:140]: We can discuss about this @tseho as you were already not a big fan of this

File: akeneo-design-system/src/components/Checkbox/Checkbox.tsx [hunk 3]
```diff
@@ -1,7 +1,21 @@
...
+import {uuid} from 'shared/uuid';
+
+const getAriaChecked = (status: CheckboxStatus): string => {
+  switch (status) {
+    case 'checked':
+      return 'true';
+    case 'unchecked':
+      return 'false';
+    case 'undetermined':
+      return 'mixed';
```
Comment (juliensnz) [inline akeneo-design-system/src/components/Checkbox/Checkbox.tsx:16]: this is the real aria naming. I don't really like it but maybe we should follow it as it's a standard... Give me your thoughts :)

File: akeneo-design-system/src/components/Checkbox/Checkbox.tsx [hunk 4]
```diff
@@ -34,7 +48,7 @@ const CheckboxContainer = styled.div<{checked: boolean; readOnly: boolean} & Ake
   width: 20px;
   border: 1px solid ${getColor('grey80')};
   border-radius: 3px;
-  outline: none;
+  overflow: hidden;
```
Comment (juliensnz) [inline akeneo-design-system/src/components/Checkbox/Checkbox.tsx:51]: to fix the focus outline

Comment (juliensnz) [reply akeneo-design-system/src/components/Checkbox/Checkbox.tsx:51]: I also think it's more accessible to have the default outline

File: akeneo-design-system/src/components/Checkbox/Checkbox.tsx [hunk 5]
```diff
@@ -72,7 +86,7 @@ const CheckboxContainer = styled.div<{checked: boolean; readOnly: boolean} & Ake
     `}
 `;
 
-const LabelContainer = styled.div<{readOnly: boolean} & AkeneoThemedProps>`
+const LabelContainer = styled.label<{readOnly: boolean} & AkeneoThemedProps>`
```
Comment (juliensnz) [inline akeneo-design-system/src/components/Checkbox/Checkbox.tsx:89]: more accessible IMHO

File: akeneo-design-system/src/components/Checkbox/Checkbox.tsx [hunk 6]
```diff
@@ -112,36 +126,56 @@ type CheckboxProps = {
...
-  if (undefined === onChange && false === readOnly) {
-    throw new Error('A Checkbox element expect an onChange attribute if not readOnly');
-  }
+const Checkbox = React.forwardRef<HTMLDivElement, CheckboxProps>(
+  (
+    {label, status, onChange, readOnly = false}: CheckboxProps,
+    forwardedRef: Ref<HTMLDivElement>
+  ): React.ReactElement => {
+    if (undefined === onChange && false === readOnly) {
+      throw new Error('A Checkbox element expect an onChange attribute if not readOnly');
```
Comment (juliensnz) [inline akeneo-design-system/src/components/Checkbox/Checkbox.tsx:135]: I am asking myself if we should use "readOnly" or "disabled". Maybe it's not important but it feels weird to have "readOnly here and "disabled" on the button

File: akeneo-design-system/src/components/Checkbox/Checkbox.tsx [hunk 7]
```diff
@@ -112,36 +126,56 @@ type CheckboxProps = {
...
+        case 'checked':
+          onChange('unchecked');
+          break;
+        case 'undetermined':
+        case 'unchecked':
+          onChange('checked');
+          break;
+      }
+    };
+    const ref = useShortcut(Key.Space, handleChange);
```
Comment (juliensnz) [inline akeneo-design-system/src/components/Checkbox/Checkbox.tsx:156]: I follow the recommended accessibility rule

File: akeneo-design-system/src/components/Checkbox/Checkbox.tsx [hunk 8]
```diff
@@ -112,36 +126,56 @@ type CheckboxProps = {
...
+    {label, status, onChange, readOnly = false}: CheckboxProps,
+    forwardedRef: Ref<HTMLDivElement>
+  ): React.ReactElement => {
+    if (undefined === onChange && false === readOnly) {
+      throw new Error('A Checkbox element expect an onChange attribute if not readOnly');
+    }
 
-  const checked = 'checked' === status;
-  const undetermined = 'undetermined' === status;
+    const [id] = useState<string>(`checkbox_${uuid()}`);
```
Comment (juliensnz) [inline akeneo-design-system/src/components/Checkbox/Checkbox.tsx:138]: tell me what you think. The idea is to have an id to use "for" and "aria-labeledby" for accessibility

File: akeneo-design-system/src/hooks/use-shortcut.ts [hunk 9]
```diff
@@ -1,16 +1,32 @@
...
+  callback: () => void,
+  externalRef: Ref<NodeType> = null
+): Ref<NodeType> => {
   const memoizedCallback = useCallback((event: KeyboardEvent) => (key === event.code ? callback() : null), [
     key,
     callback,
   ]);
 
+  const internalRef = useRef<NodeType>(null);
+  const ref = null === externalRef ? internalRef : externalRef;
```
Comment (juliensnz) [inline akeneo-design-system/src/hooks/use-shortcut.ts:15]: maybe I should add a small description here to explain more. The idea is to use an external ref if provided or use a generated ref to return it

File: akeneo-design-system/src/shared/key.ts [hunk 10]
```diff
@@ -1,5 +1,5 @@
 enum Key {
-  Space = ' ',
+  Space = 'Space',
```
Comment (juliensnz) [inline akeneo-design-system/src/shared/key.ts:2]: I don't know why this changed but it was not working with the " "

File: akeneo-design-system/src/shared/uuid.ts [hunk 11]
```diff
@@ -0,0 +1,15 @@
+const uuid = (): string => {
```
Comment (juliensnz) [inline akeneo-design-system/src/shared/uuid.ts:1]: I don't want to import a 113 kB dependency to generate uuids so I copy pasted this snippet from mdn

File: akeneo-design-system/src/components/Checkbox/Checkbox.tsx [hunk 12]
```diff
@@ -1,7 +1,21 @@
-import React from 'react';
+import React, {Ref, useState} from 'react';
 import styled, {css, keyframes} from 'styled-components';
 import {AkeneoThemedProps, getColor} from 'theme';
 import {CheckIcon, PartialCheckIcon} from 'icons';
+import {useShortcut} from 'hooks/use-shortcut';
```
Comment (ValentinMumble) [inline akeneo-design-system/src/components/Checkbox/Checkbox.tsx:5]:
```suggestion
import {useShortcut} from 'hooks';
```
And add an `index.ts` in the hooks folder

File: akeneo-design-system/src/components/Checkbox/Checkbox.tsx [hunk 13]
```diff
@@ -1,7 +1,21 @@
...
+import {Key} from 'shared/key';
```
Comment (ValentinMumble) [inline akeneo-design-system/src/components/Checkbox/Checkbox.tsx:6]:
```suggestion
import {Key, uuid} from 'shared';
```
Same, with an `index.ts` in shared folder

File: akeneo-design-system/src/hooks/use-shortcut.ts [hunk 14]
```diff
@@ -1,6 +1,47 @@
-import {useCallback, useEffect} from 'react';
+// import {Ref, useCallback, useEffect, useRef} from 'react';
```
Comment (ValentinMumble) [inline akeneo-design-system/src/hooks/use-shortcut.ts:1]: We should also rename this file to `useShorcut.ts` to be consistent?

Comment (juliensnz) [reply akeneo-design-system/src/hooks/use-shortcut.ts:1]: Yes, we could

File: akeneo-design-system/src/components/Checkbox/Checkbox.stories.mdx [hunk 15]
```diff
@@ -48,12 +46,12 @@ Use this playground to test the checkbox component
 <Canvas>
   <Story name="Standard">
     {(args) => {
-      const [{status}, updateArgs] = useArgs();
-      const nextStatus = 'checked' === status ? 'unchecked' : 'checked';
+      const [{checked}, updateArgs] = useArgs();
+      const nextChecked = 'chectrueked' === checked ? 'false' : 'true';
```
Comment (ValentinMumble) [inline akeneo-design-system/src/components/Checkbox/Checkbox.stories.mdx:50]:
```suggestion
      const nextChecked = 'checked' === checked ? 'false' : 'true';
```

File: akeneo-design-system/src/components/Checkbox/Checkbox.tsx [hunk 16]
```diff
@@ -85,63 +87,87 @@ const LabelContainer = styled.div<{readOnly: boolean} & AkeneoThemedProps>`
...
    */
-  label?: string;
+  onChange?: (value: CheckboxChecked) => void;
 
   /**
-   * The handler called when clicking on Checkbox.
+   * Label of the checkbox.
    */
-  onChange?: (value: CheckboxStatus) => void;
+  children: string | ReactNode;
```
Comment (StevenVAIDIE) [inline akeneo-design-system/src/components/Checkbox/Checkbox.tsx:111]: I think that children can be null

Comment (ValentinMumble) [inline akeneo-design-system/src/components/Checkbox/Checkbox.tsx:111]:
```suggestion
  children?: string;
```

**Review by SamirBoulil — APPROVED** [review]

File: akeneo-design-system/src/hooks/useShortcut.ts [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+
+  const internalRef = useRef<NodeType>(null);
+  const ref = null === externalRef ? internalRef : externalRef;
+
+  useEffect(() => {
+    if (typeof ref !== 'function' && null !== ref.current) {
+      const element = (ref.current as unknown) as NodeType;
+
+      element.addEventListener('keydown', memoizedCallback);
+      return () => element.removeEventListener('keydown', memoizedCallback);
```
Comment (ValentinMumble) [inline akeneo-design-system/src/hooks/useShortcut.ts:29]: Code coverage doesn't whine on this part?

**Review by ValentinMumble — APPROVED** [review]

File: akeneo-design-system/src/components/Checkbox/Checkbox.stories.mdx [hunk 18]
```diff
@@ -9,13 +9,11 @@ import {useArgs, updateArgs} from '@storybook/client-api';
...
-    status: {type: {control: {type: 'select', options: ['checked', 'unchecked', 'undetermined']}}},
-    label: {control: {type: 'text'}},
-    onChange: {action: 'Checkbox component onChange'}
+    checked: {type: {control: {type: 'select', options: ['true', 'false', 'mixed']}}},
+    onChange: {action: 'Checkbox component onChange'},
   }}
   args={{
-    label: 'Checkbox',
-    status: 'checked'
+    checked: 'true'
```
Comment (StevenVAIDIE) [inline akeneo-design-system/src/components/Checkbox/Checkbox.stories.mdx:16]: In playgroup children is not dynamic 

```suggestion
    checked: 'true',
    children: 'Checkbox'
```

And remove 'Checkbox' children in https://github.com/akeneo/pim-community-dev/pull/12811/files#diff-0e0c27478afb0bc422e79163d861f955R54
https://github.com/akeneo/pim-community-dev/pull/12811/files#diff-0e0c27478afb0bc422e79163d861f955R54

File: akeneo-design-system/src/components/Checkbox/Checkbox.tsx (hunk 16 above)
Comment (StevenVAIDIE) [inline akeneo-design-system/src/components/Checkbox/Checkbox.tsx:111]: I think it can be null

**Review by StevenVAIDIE — APPROVED** [review]

File: akeneo-design-system/src/hooks/useShortcut.ts (hunk 17 above)
Comment (juliensnz) [reply akeneo-design-system/src/hooks/useShortcut.ts:29]: nope

File: akeneo-design-system/src/components/Checkbox/Checkbox.tsx (hunk 16 above)
Comment (juliensnz) [reply akeneo-design-system/src/components/Checkbox/Checkbox.tsx]: I allow more here

**Pull request merged**